src/NexusMods.App.UI/WorkspaceSystem/GridUtils.cs

```diff
@@ -438,94 +438,245 @@ internal static bool IsPerfectGrid(WorkspaceGridState gridState)
         return res.UnionById(updates);
     }
 
-    internal static IReadOnlyList<ResizerInfo> GetResizers(
-        ImmutableDictionary<PanelId, Rect> currentState,
-        bool isWorkspaceHorizontal = true)
+    internal static List<ResizerInfo> GetResizers2(WorkspaceGridState workspaceState)
     {
-        var tmp = new List<ResizerInfo>(capacity: currentState.Count);
-        var adjacentPanels = new List<ResizerInfo>(capacity: currentState.Count);
+        return workspaceState.IsHorizontal
+            ? GetResizersForHorizontal(workspaceState)
+            : GetResizersForVertical(workspaceState);
+    }
+
+    private static List<ResizerInfo> GetResizersForVertical(WorkspaceGridState currentState)
+    {
+        var res = new List<ResizerInfo>();
 
-        // Step 1: fill the tmp List with all Resizers
-        foreach (var current in currentState)
+        var (rowCount, maxColumnCount) = currentState.CountRows();
+
+        Span<WorkspaceGridState.RowInfo> seenRows = stackalloc WorkspaceGridState.RowInfo[rowCount];
+        using var rowEnumerator = new WorkspaceGridState.RowEnumerator(currentState, seenRows);
+
+        Span<PanelGridState> columnBuffer = stackalloc PanelGridState[maxColumnCount];
+        Span<PanelGridState> columnsToAdd = stackalloc PanelGridState[maxColumnCount];
+
+        while (rowEnumerator.MoveNext(columnBuffer))
         {
-            var currentRect = current.Value;
+            var numColumnsToAdd = 0;
+
+            var row = rowEnumerator.Current;
+            var info = row.Info;
+            var columns = row.Columns;
+
+            for (var i = 0; i < columns.Length - 1; i++)
+            {
+                var (curId, curRect) = columns[i];
+                var (nextId, nextRect) = columns[i + 1];
+
+                var (columnStart, columnEnd) = MathUtils.GetResizerPoints(curRect, nextRect, WorkspaceGridState.AdjacencyKind.SameRow);
+                if (!HasOther(columnStart, columnEnd, columns))
+                {
+                    res.Add(new ResizerInfo(columnStart, columnEnd, IsHorizontal: false, [curId, nextId]));
+                }
+            }
+
+            if (rowCount == 1) break;
+
+            var (minX, maxX) = (1.0, 0.0);
+            foreach (var column in columns)
+            {
+                var rect = column.Rect;
+
+                // panels that go across rows are skipped
+                if (column.IsCrossRow(info)) continue;
+
+                minX = Math.Min(rect.Left, minX);
+                maxX = Math.Max(rect.Right, maxX);
+
+                columnsToAdd[numColumnsToAdd++] = column;
+            }
 
-            // Step 1.1: go through all panels and find their adjacent panels
-            foreach (var other in currentState)
+            if (minX > maxX)
             {
-                if (other.Key == current.Key) continue;
+                Debugger.Break();
```

Comment:
Do you want to leave this in, or is it a leftover from debugging?

Reply:
I'm not sure. The condition `minX > maxX` should be impossible, but our code depends on that assumption to be correct. I think I'll use a `Debug.Assert` for this.